### PR TITLE
[Feature][STACK-2017][vThunder support] - a10-octavia vThunder flow support

### PR DESCRIPTION
**Pull request by ytsai-a10**
## Description
Before a10-octavia support vThunder flows, which a10-octavia can automatically create vThunder instance when loadbalancer is created in a openstack project. But the flow didn’t be maintained for a while. For customer requirements a10-octavia v1.2 will support vThunder as amphora with most features. 

Supported features:
- basic functionality
- VRRP-A floating ip support
- Flavor support
- SNAT
- write-memory enhancement

Design document: 
https://teams.microsoft.com/l/file/96CCF61A-DE50-4BEB-9B14-6A4D215270C3?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.2%2FvThunder%20support%2F%5BDD%5D%20vThunder%20support.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2017
https://a10networks.atlassian.net/browse/STACK-2016

## Technical Approach
- Move features in hardware thunder flow to vThunder flow

## Config Changes
Please reference following document for vThunder configuration and setup:
https://teams.microsoft.com/l/file/CED51F5A-0A31-4B09-8789-934939826F58?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.2%2FvThunder%20support%2Fvthunder%20setup.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

## Test Cases
Tested Features:
- basic functionality (tested by automation)
- VRRP-A floating ip support (tested by Adeeb)
- Flavor support (tested by Neha)
- SNAT (tested by Neha)
- write-memory enhancement (tested by Adeeb)

## Manual Testing
Please check PRs for detail test items and results:
- basic operations: https://github.com/a10networks/a10-octavia/pull/322
- VCS and Active-Standby: https://github.com/a10networks/a10-octavia/pull/327
- VRID and VRRP-A Floating-IP support: https://github.com/a10networks/a10-octavia/pull/335
- Flavor Support: https://github.com/a10networks/a10-octavia/pull/323
- SNAT: https://github.com/a10networks/a10-octavia/pull/321
- Periodically Write Memory: https://github.com/a10networks/a10-octavia/pull/325
- Periodically Write Memory for Active-Standby topology: https://github.com/a10networks/a10-octavia/pull/330

### CODE REVIEW

File: a10_octavia/common/a10constants.py [hunk 1]
```diff
@@ -93,6 +93,7 @@
 CONFIRM_VRRP_STATUS = 'confirm-vrrp-status'
 ALLOCATE_VIP = 'allocate-vip'
 UPDATE_VIP_AFTER_ALLOCATION = 'update-vip-after-allocation'
+GET_COMPUTE_FOR_PROJECT = 'get_compute_for_project'
```
Comment (hthompson-a10) [inline a10_octavia/common/a10constants.py:96]: Need to turn underscores '_' to dashes '-'

File: a10_octavia/controller/worker/controller_worker.py [hunk 2]
```diff
@@ -903,9 +903,13 @@ def perform_write_memory(self, thunders):
         store = {a10constants.WRITE_MEM_SHARED_PART: True}
 
         for vthunder in thunders:
+            deleteCompute = False
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/controller_worker.py:906]: We use underscores for vars not camel case!

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 3]
```diff
@@ -204,10 +210,21 @@ def get_delete_load_balancer_flow(self, lb, deleteCompute):
...
+        if not deleteCompute:
+            delete_LB_flow.add(vthunder_tasks.WriteMemory(
+                requires=a10constants.VTHUNDER))
         delete_LB_flow.add(a10_database_tasks.SetThunderUpdatedAt(
             requires=a10constants.VTHUNDER))
+        delete_LB_flow.add(a10_database_tasks.GetBackupVThunderByLoadBalancer(
+            requires=constants.LOADBALANCER,
+            provides=a10constants.BACKUP_VTHUNDER))
+        delete_LB_flow.add(a10_database_tasks.MarkVThunderStatusInDB(
+            name="DELETED_BACKUP",
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:222]: This name should map to a constant which follows our syntax

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py [hunk 4]
```diff
@@ -204,10 +210,21 @@ def get_delete_load_balancer_flow(self, lb, deleteCompute):
...
+            rebind={a10constants.VTHUNDER: a10constants.BACKUP_VTHUNDER},
+            inject={"status": constants.DELETED}))
+        delete_LB_flow.add(a10_database_tasks.SetThunderUpdatedAt(
+            name="SET_BACKUP_UPDATEDAT",
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:226]: This name should map to a constant which follows our syntax

File: a10_octavia/controller/worker/flows/a10_member_flows.py [hunk 5]
```diff
@@ -96,9 +97,24 @@ def get_create_member_flow(self, topology):
         create_member_flow.add(a10_database_tasks.CountMembersWithIP(
             requires=constants.MEMBER, provides=a10constants.MEMBER_COUNT_IP
         ))
+        create_member_flow.add(a10_database_tasks.GetFlavorData(
+            rebind={a10constants.LB_RESOURCE: constants.LOADBALANCER},
+            provides=constants.FLAVOR))
+        create_member_flow.add(server_tasks.MemberFindNatPool(
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/flows/a10_member_flows.py:103]: Let's move all the NatPool flow tasks into a subflow to cut down on LOC and make it more manageable

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/common/a10constants.py (hunk 1 above)
Comment (ytsai-a10) [reply a10_octavia/common/a10constants.py:96]: done

File: a10_octavia/controller/worker/controller_worker.py (hunk 2 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/controller_worker.py:906]: done

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py (hunk 3 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:222]: done

File: a10_octavia/controller/worker/flows/a10_load_balancer_flows.py (hunk 4 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/a10_load_balancer_flows.py:226]: done

File: a10_octavia/controller/worker/flows/a10_member_flows.py (hunk 5 above)
Comment (ytsai-a10) [reply a10_octavia/controller/worker/flows/a10_member_flows.py]: done

**Review by rahulkatre-a10 — APPROVED** [review]

**Review by hthompson-a10 — APPROVED** [review]

**Pull request merged**